main_board/app/src/optics/ir_camera_system/ir_camera_timer_settings.c

```diff
@@ -10,97 +10,49 @@
 LOG_MODULE_REGISTER(ir_camera_timer_settings,
                     CONFIG_IR_CAMERA_SYSTEM_LOG_LEVEL);
 
-#define MAX_PSC_DIV 65536U
-
 ret_code_t
 timer_settings_from_on_time_us(
     uint16_t on_time_us,
     const struct ir_camera_timer_settings *current_settings,
     struct ir_camera_timer_settings *new_settings)
 {
     ret_code_t ret;
-    struct ir_camera_timer_settings ts = {0};
+    struct ir_camera_timer_settings ts = *current_settings;
 
     if (on_time_us == 0) {
         ret = RET_SUCCESS;
-        ts = *current_settings;
         ts.ccr = 0;
         ts.on_time_in_us = 0;
     } else if (on_time_us > IR_CAMERA_SYSTEM_MAX_IR_LED_ON_TIME_US) {
         ret = RET_ERROR_INVALID_PARAM;
     } else if (current_settings->fps == 0) {
+        // save on-time duration for when FPS is set, but dont calculate ccr
+        // TODO: when IR LED timers will be independent form
+        // CAMERA_TRIGGER_TIMER this case can be handled in the default case
         ret = RET_SUCCESS;
-        ts = *current_settings;
         ts.on_time_in_us = on_time_us;
+    } else if ((1e6 / current_settings->fps *
+                IR_CAMERA_SYSTEM_MAX_IR_LED_DUTY_CYCLE) < on_time_us) {
+        ret = RET_ERROR_INVALID_PARAM;
+        LOG_ERR(
+            "On-time duration must not exceed %uµs for the current FPS setting",
+            (uint16_t)(1e6 / current_settings->fps *
+                       IR_CAMERA_SYSTEM_MAX_IR_LED_DUTY_CYCLE));
     } else {
-        ts = *current_settings;
-
-        // on-time duration must not be longer than
-        // IR_CAMERA_SYSTEM_MAX_IR_LED_ON_TIME_US
+        ret = RET_SUCCESS;
         ts.on_time_in_us = on_time_us;
-
-        // find the FPS that matches the on-time duration
-        ts.fps = 100000.0 / ts.on_time_in_us;
-
-        ts.psc = ASSUMED_TIMER_CLOCK_FREQ / (MAX_PSC_DIV * ts.fps);
-        ts.arr = ASSUMED_TIMER_CLOCK_FREQ / ((ts.psc + 1) * ts.fps);
+        uint32_t accuracy_us = (ts.psc + 1) / ASSUMED_TIMER_CLOCK_FREQ_MHZ;
+        if (accuracy_us > 1) {
+            LOG_WRN("on-time duration accuracy: %uus", accuracy_us);
+        }
+        // calculate the ccr value for the new on-time duration
         ts.ccr =
             (ASSUMED_TIMER_CLOCK_FREQ_MHZ * ts.on_time_in_us) / (ts.psc + 1);
```

Comment:
Add typecast to uint32_t. `ts.on_time_in_us` is uint16 with a max value of 8000, so it can overflow if multiplied by 170.